File: ITP_KANDY_7-main/backend/src/infrastructure/models/Purchase.js
/**
 * Infrastructure Layer: MongoDB Schema for Purchase Records.
 * manages the procurement ledger for stock replenishment and supplier debt tracking.
 */

const mongoose = require('mongoose');

/**
 * Sub-Schema: Procurement Line Item.
 * Tracks specific products added to inventory during a restock event.
 */
const PurchaseItemSchema = new mongoose.Schema({
    productId: { type: String, required: true }, // Source product identity
    productName: { type: String, required: true, alias: 'name' }, // Flat name for record stability
    quantity: { type: Number, required: true }, // Volume added to stock
    costPrice: { type: Number, required: true, alias: 'price' }, // Per-unit cost to the merchant
    subtotal: { type: Number, required: true } // Item-level cost (Qty * CostPrice)
}, { _id: false });

/**
 * Main Schema: Purchase/Inventory Ledger.
 */
const PurchaseSchema = new mongoose.Schema({
    _id: { type: String, required: true }, // Unique Transaction ID
    ownerId: { type: String, required: true }, // Multi-tenant key (indexed via compound index below)
    supplierId: { type: String, required: true }, // Linked supply partner
    supplierName: { type: String, default: '' }, // Denormalized name for display
    invoiceNumber: { type: String, default: '' }, // Supplier's physical invoice reference
    purchaseDate: { type: String, index: true }, // User-selected transaction date
    items: [PurchaseItemSchema], // Embedded array of items restocked
    subtotal: { type: Number, default: 0 }, // Sum of items before tax/discounts
    tax: { type: Number, default: 0 }, // Additional procurement costs
    totalAmount: { type: Number, required: true }, // Grand total cost of the purchase
    amountPaid: { type: Number, default: 0 }, // Cash already settled with supplier
    remaining: { type: Number, default: 0 }, // Unpaid balance (Debt)
    status: { type: String, default: 'pending' }, // e.g., 'paid', 'partial', 'pending'
    paymentMethod: { type: String, default: 'cash' }, // Settlement channel
    notes: { type: String, default: '' }, // Optional merchant annotations
    createdAt: { type: String, index: true }, // Audit: Entry generation date
    updatedAt: { type: String } // Last modified audit time
}, {
    _id: false,
    timestamps: false
});

// Performance: Indexes for multi-tenant chronological reporting.
PurchaseSchema.index({ ownerId: 1, createdAt: -1 });
PurchaseSchema.index({ ownerId: 1, purchaseDate: -1 });

/**
 * Logic: Interface Compatibility.
 * Maps MongoDB '_id' to the domain-standard 'id' field.
 */
PurchaseSchema.virtual('id').get(function() {
    return this._id;
});

PurchaseSchema.set('toJSON', { virtuals: true });

// Module Export: Purchase data model for inventory logistics.
module.exports = mongoose.model('Purchase', PurchaseSchema);

File: ITP_KANDY_7-main/frontend/lib/features/suppliers/domain/entities/purchase.dart
// ------------------------------------------------------------------------------
// File: purchase.dart
// Purpose: Domain entity representing a single inventory restock transaction.
// Rationale: Models the full lifecycle of a supplier purchase including line
//   items, tax, payment status, and residual debt. Used across Supplier
//   module for purchase recording, settlement, and PDF receipt generation.
// ------------------------------------------------------------------------------
class Purchase {
  final String id;
  final String supplierId;
  final String supplierName;
  final String invoiceNumber;
  final String purchaseDate;
  final List<dynamic> items;
  final double subtotal;
  final double tax;
  final double totalAmount;
  final double amountPaid;
  final double remaining;
  final String status;
  final String paymentMethod;
  final String notes;
  final String createdAt;

  Purchase({
    required this.id,
    this.supplierId = '',
    this.supplierName = '',
    this.invoiceNumber = '',
    this.purchaseDate = '',
    this.items = const [], // The list of products being restocked
    this.subtotal = 0,
    this.tax = 0,
    this.totalAmount = 0, // The final price the shop owner paid to the supplier
    this.amountPaid = 0, // How much of the total has already been paid
    this.remaining = 0, // The leftover debt to the supplier for this purchase
    this.status = 'pending',
    this.paymentMethod = 'cash',
    this.notes = '',
    this.createdAt = '',
  });

  factory Purchase.fromJson(Map<String, dynamic> json) {
    return Purchase(
      id: json['id'] ?? '',
      supplierId: json['supplierId'] ?? '',
      supplierName: json['supplierName'] ?? '',
      invoiceNumber: json['invoiceNumber'] ?? '',
      purchaseDate: json['purchaseDate'] ?? '',
      items: json['items'] ?? [],
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      tax: (json['tax'] ?? 0).toDouble(),
      totalAmount: (json['totalAmount'] ?? 0).toDouble(),
      amountPaid: (json['amountPaid'] ?? 0).toDouble(),
      remaining: (json['remaining'] ?? 0).toDouble(),
      status: json['status'] ?? 'pending',
      paymentMethod: json['paymentMethod'] ?? 'cash',
      notes: json['notes'] ?? '',
      createdAt: json['createdAt'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'supplierId': supplierId,
      'supplierName': supplierName,
      'invoiceNumber': invoiceNumber,
      'purchaseDate': purchaseDate,
      'items': items,
      'subtotal': subtotal,
      'tax': tax,
      'totalAmount': totalAmount,
      'amountPaid': amountPaid,
      'status': status,
      'paymentMethod': paymentMethod,
      'notes': notes,
    };
  }
}

